src/System.Text.Http.Parser/HttpParser.cs

```diff
@@ -340,6 +340,94 @@ public class HttpParser : IHttpParser
             }
         }
 
+        public unsafe bool ParseHeaders<T>(T handler, ReadOnlyBytes buffer, out int consumedBytes) where T : IHttpHeadersHandler
+        {
+            var span = buffer.First.Span;
+            var rest = buffer.Rest;
+            var length = span.Length;
+            int index = -1;
+            State state = State.ParsingName;
+            int nameStart = 0, nameEnd = 0, valueStart = 0, valueEnd = 0;
+            bool sawCRLF = false;
+            bool straddling = false;
+            fixed (byte* pBuffer = &span.DangerousGetPinnableReference()) {
+                while (true) {
+                    index++;
+                    // get next byte 
+                    while (index == length) {
```

Comment:
But now, I noticed there is a bug here. I initally had index starting at 0 then I changed it to -1, and now the check does not work. I will fix.